readthedocs/oauth/migrate.py

```diff
@@ -0,0 +1,358 @@
+
+
+@dataclass
+class InstallationTargetGroup:
+    """Group of repositories that should be installed in the same target (user or organization)."""
+
+    target_id: int
+    target_type: GitHubAccountType
+    target_name: str
+    repository_ids: set[int]
+
+    @property
+    def link(self):
+        """
+        Create a link to install the GitHub App on the target with the required repositories pre-selected.
+
+        See https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/migrating-oauth-apps-to-github-apps#prompt-users-to-install-your-github-app.
+        """
+        repository_ids = []
+        for repository_id in self.repository_ids:
+            repository_ids.append(f"&repository_ids[]={repository_id}")
+        repository_ids = "".join(repository_ids)
+
+        base_url = (
+            f"https://github.com/apps/{settings.GITHUB_APP_NAME}/installations/new/permissions"
+        )
+        return f"{base_url}?suggested_target_id={self.target_id}{repository_ids}"
+
+    @property
+    def installed(self):
+        """
+        Check if the app has been in all required repositories.
+
+        The the list of repository_ids is not empty, it means that the app still needs to be installed in some repositories,
+        or that the app hasn't been installed at all in the target account.
+        """
+        return not bool(self.repository_ids)
+
+
+@dataclass
+class MigrationTarget:
+    """Information about an individual project that needs to be migrated."""
+
+    project: Project
+    has_installation: bool
+    is_admin: bool
+    target_id: int
+
+    @property
+    def installation_link(self):
+        """
+        Create a link to install the GitHub App on the target repository.
+
+        See https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/migrating-oauth-apps-to-github-apps
+        """
+        base_url = (
+            f"https://github.com/apps/{settings.GITHUB_APP_NAME}/installations/new/permissions"
+        )
+        return f"{base_url}?suggested_target_id={self.target_id}&repository_ids[]={self.project.remote_repository.remote_id}"
+
+    @property
+    def can_be_migrated(self):
+        """
+        Check if the project can be migrated.
+
+        The project can be migrated if the user is an admin on the repository and the GitHub App is installed.
+        """
+        return self.is_admin and self.has_installation
+
+
+@dataclass
+class MigrationResult:
+    """Result of a migration operation."""
+
+    webhook_removed: bool
+    ssh_key_removed: bool
+
+
+class MigrationError(Exception):
+    """Error raised when a migration operation fails."""
+
+    pass
+
+
+def get_installation_target_groups_for_user(user) -> list[InstallationTargetGroup]:
+    """Get all targets (accounts and organizations) that the user needs to install the GitHub App on."""
+    # Since we don't save the ID of the owner of each repository, we group all repositories
```

Comment:
I read this sentence a few times now: "Since we don't save the ID of the owner of each repository". Would it be useful to start saving the ID of the owner?

Reply:
Yeah, maybe something like owner, maybe see if it makes sense to merge it with the organization like GitHub does (owners can be of type user or organization)

Reply:
https://github.com/readthedocs/readthedocs.org/issues/12116